cmd/gateway/main.go

```diff
@@ -0,0 +1,167 @@
+	certBytes, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+	block, _ := pem.Decode(certBytes)
+	return x509.ParseCertificates(block.Bytes)
+}
+
+type commandlineOpts struct {
+	URL      string `long:"url" required:"true" description:"AM URL"`
+	Realm    string `long:"realm" required:"true" description:"AM Realm"`
+	Tree     string `long:"tree" required:"true" description:"Authentication tree"`
+	Name     string `long:"name" required:"true" description:"Gateway name"`
+	Address  string `long:"address" required:"true" description:"CoAP Address of Gateway"`
+	KeyFile  string `long:"key" required:"true" description:"The file containing the Gateway's confirmation key"`
+	KeyID    string `long:"kid" description:"The Gateway's confirmation key ID"`
+	CertFile string `long:"cert" description:"The file containing the Gateway's certificate"`
+	// see time.ParseDuration for valid timeout strings
+	Timeout time.Duration `long:"timeout" default:"5s" description:"Timeout for AM communications"`
+	Debug   bool          `short:"d" long:"debug" description:"Switch on debug"`
+}
+
+func (o commandlineOpts) String() string {
+	return fmt.Sprintf(
+		`commandline options
+	url: %s
+	realm: %s
+	tree: %s
+	name: %s
+	address: %s
+	key: %s
+	kid: %s
+	certificate: %s
+	timeout %v
+	debug: %v`,
+		o.URL, o.Realm, o.Tree, o.Name, o.Address, o.KeyFile, o.KeyID, o.CertFile, o.Timeout, o.Debug)
+}
+
+// runGateway initialises and runs a Thing Gateway
+func runGateway() error {
+	var opts commandlineOpts
+	_, err := flags.Parse(&opts)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("%v\n", opts)
+
+	if opts.Debug {
+		// pipe debug to standard out
+		thing.DebugLogger.SetOutput(os.Stdout)
+	}
+
+	amKey, err := loadKey(opts.KeyFile)
+	if err != nil {
+		return err
+	}
+
+	if opts.KeyID == "" {
+		opts.KeyID, err = thing.JWKThumbprint(amKey)
+		if err != nil {
+			return err
+		}
+	}
+
+	callbacks := []callback.Handler{
+		callback.AuthenticateHandler{
+			Realm:   opts.Realm,
+			ThingID: opts.Name,
+			KeyID:   opts.KeyID,
+			Key:     amKey,
+		}}
+	if opts.CertFile != "" {
+		certs, err := loadCertificates(opts.CertFile)
+		if err != nil {
+			return err
+		}
+		callbacks = append(callbacks, callback.RegisterHandler{
+			Realm:        opts.Realm,
+			ThingID:      opts.Name,
+			ThingType:    callback.TypeGateway,
+			KeyID:        opts.KeyID,
+			Key:          amKey,
+			Certificates: certs,
+		})
+
+	}
+	gateway := thing.NewThingGateway(opts.URL, opts.Realm, opts.Tree, opts.Timeout, callbacks)
+
+	err = gateway.Initialise()
+	if err != nil {
+		return err
+	}
+
+	serverKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return err
+	}
+	err = gateway.StartCOAPServer(opts.Address, serverKey)
+	if err != nil {
+		return err
+	}
+	defer gateway.ShutdownCOAPServer()
+
+	fmt.Println("Thing Gateway server started. Press a key to exit.")
+	bufio.NewScanner(os.Stdin).Scan()
```

Comment:
If the gateway is started in the background then this might not work. Maybe we should use `os.Signal`(https://gobyexample.com/signals) so we can respond to OS commands instead.